source/common/common/regex.cc

```diff
@@ -47,12 +49,54 @@ class CompiledGoogleReMatcher : public CompiledMatcher {
       throw EnvoyException(regex_.error());
     }
 
-    const uint32_t max_program_size =
-        PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.google_re2(), max_program_size, 100);
-    if (static_cast<uint32_t>(regex_.ProgramSize()) > max_program_size) {
-      throw EnvoyException(fmt::format("regex '{}' RE2 program size of {} > max program size of "
-                                       "{}. Increase configured max program size if necessary.",
-                                       config.regex(), regex_.ProgramSize(), max_program_size));
+    const uint32_t regex_program_size = static_cast<uint32_t>(regex_.ProgramSize());
+
+    // Check if the deprecated field max_program_size is set first, and follow the old logic if so.
+    if (config.google_re2().has_max_program_size()) {
+      const uint32_t max_program_size =
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.google_re2(), max_program_size, 100);
+      if (static_cast<uint32_t>(regex_program_size > max_program_size)) {
+        throw EnvoyException(fmt::format("regex '{}' RE2 program size of {} > max program size of "
+                                         "{}. Increase configured max program size if necessary.",
+                                         config.regex(), regex_program_size, max_program_size));
+      }
+      return;
+    }
+
+    Runtime::Loader* runtime = Runtime::LoaderSingleton::getExisting();
+    if (runtime) {
+      Stats::Scope& root_scope = runtime->getRootScope();
+
+      Stats::StatNameManagedStorage program_size_stat_name("re2.max_program_size",
```

Comment:
it would be better to capture this histogram in a stats object created at startup, rather than taking a global symbol table lock on every regex compile, particularly as you are creating this stat off the root scope. Just do it once.

Reply:
ping on this (& below)

Reply:
ack - I'm still trying to understand where it should be created here since the match method creates a matcher each time it's called unless I'm misunderstanding something

Reply:
I thought about this a bit; given that we have Runtime::getRootScope() we could put some global shared stats here too, and add Runtime::getRootStats()

I personally would prefer better plumbing of contexts through constructors for this rather than using singletons but given you are already getting root_scope from thin air we might as well get root_stats as well.

Having said that, it looks like getRootScope() was added in this PR and it doesn't look at all related to the other stuff in Runtime.

Reply:
Ah I see I was already pinged about this a month ago and didn't object then, so I guess I shouldn't object now :) And I do agree this makes this change a lot more surgical.

Reply:
Yeah my 2c is this is not a high perf path and it's not worth the additional complexity to pre-allocate this.

Reply:
OK, can we leave a TODO(perf) here in case this comes up?

Thanks!